### PR TITLE
[SYCLomatic] Impl a framework to migrate math functions

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/APINamesMathRewrite.inc [hunk 1]
```diff
@@ -6,28 +6,23 @@
...
-                IsDirectCalleeHasAttribute<CUDADeviceAttr>(),
-                makeCheckNot(IsDirectCalleeHasAttribute<CUDAHostAttr>()))),
-        makeCheckOr(IsContextCallHasAttribute<CUDADeviceAttr>(),
-                    IsContextCallHasAttribute<CUDAGlobalAttr>())),
-    CALL_FACTORY_ENTRY("rsqrt",
-                       CALL(MapNames::getClNamespace(false, true) + "rsqrt",
-                            CAST_IF_NEED(makeLiteral("double"), ARG(0)))),
-    NO_REWRITE_FUNCNAME_FACTORY_ENTRY("rsqrt", "rsqrt"))
+MATH_API_REWRITER_HOST_DEVICE(
+    MATH_API_REWRITER_HOST_2(
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINamesMathRewrite.inc]: Consider to remove _2?  not easy to understand.

File: clang/lib/DPCT/CallExprRewriterCommon.h [hunk 2]
```diff
@@ -59,21 +59,6 @@ inline bool isArgMigratedToAccessor(const CallExpr *Call, unsigned Index) {
   return false;
 }
 
-inline std::string getTypecastName(const CallExpr *Call) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CallExprRewriterCommon.h]: For code clean , please use seperate PR.

File: clang/lib/DPCT/CallExprRewriterCommon.h [hunk 3]
```diff
@@ -1436,15 +1415,30 @@ class CheckDerefedTypeBeforeCast {
...
   int value;
   int index;
-
 public:
   CheckArgIsConstantIntWithValue(int idx, int val) : value(val), index(idx) {}
   bool operator()(const CallExpr *C) {
-    auto Arg = C->getArg(index);
-    Expr::EvalResult Result;
+    auto Arg = C->getArg(index)->IgnoreImpCasts();
+    Expr::EvalResult ResultInt;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CallExprRewriterCommon.h]: please add some commments to tell show case the updated feature.

File: clang/lib/DPCT/CallExprRewriterMath.cpp [hunk 4]
```diff
@@ -1203,6 +1205,275 @@ Optional<std::string> MathBinaryOperatorRewriter::rewrite() {
...
+    return SEA.hasSideEffects();
+  }
+};
+}
+
+inline std::pair<std::string, std::shared_ptr<CallExprRewriterFactoryBase>>
+createMathAPIRewriterDeviceImpl(
+    const std::string &NAME, std::function<bool(const CallExpr *)> PerfPred,
+    std::pair<std::string, std::shared_ptr<CallExprRewriterFactoryBase>>
+        DEVICE_PERF,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CallExprRewriterMath.cpp]: DevicePerf

Comment (zhimingwang36) [reply clang/lib/DPCT/CallExprRewriterMath.cpp]: NAME ==> Name

File: clang/lib/DPCT/APINamesMathRewrite.inc [hunk 5]
```diff
@@ -6,28 +6,23 @@
...
-    NO_REWRITE_FUNCNAME_FACTORY_ENTRY("rsqrt", "rsqrt"))
+MATH_API_REWRITER_HOST_DEVICE(
+    MATH_API_REWRITER_HOST(
+        "rsqrt",
+        CALL_FACTORY_ENTRY(
+            "rsqrt", CALL(MapNames::getClNamespace(false, true) + "rsqrt",
+                          ARG(0)))),
+    MATH_API_REWRITER_DEVICE(
+        "rsqrt",
+        {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APINamesMathRewrite.inc]: Line 17 ~22 is a little diffcult to understand,  can you improve it?

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**